src/connecpy/_protocol.py

```diff
@@ -0,0 +1,106 @@
+from dataclasses import asdict, dataclass
+from http import HTTPStatus
+import json
+
+from .errors import Errors
+from .exceptions import ConnecpyServerException
+
+
+# Define a custom class for HTTP Status to allow adding 499 status code
+@dataclass(frozen=True)
+class ExtendedHTTPStatus:
+    code: int
+    reason: str
+
+    @staticmethod
+    def from_http_status(status: HTTPStatus) -> "ExtendedHTTPStatus":
+        return ExtendedHTTPStatus(code=status.value, reason=status.phrase)
+
+
+# Dedupe statuses that are mapped multiple times
+_BAD_REQUEST = ExtendedHTTPStatus.from_http_status(HTTPStatus.BAD_REQUEST)
+_CONFLICT = ExtendedHTTPStatus.from_http_status(HTTPStatus.CONFLICT)
+_INTERNAL_SERVER_ERROR = ExtendedHTTPStatus.from_http_status(
+    HTTPStatus.INTERNAL_SERVER_ERROR
+)
+
+_error_to_http_status = {
+    Errors.Canceled: ExtendedHTTPStatus(499, "Client Closed Request"),
+    Errors.Unknown: _INTERNAL_SERVER_ERROR,
+    Errors.InvalidArgument: _BAD_REQUEST,
+    Errors.DeadlineExceeded: ExtendedHTTPStatus.from_http_status(
+        HTTPStatus.GATEWAY_TIMEOUT
+    ),
+    Errors.NotFound: ExtendedHTTPStatus.from_http_status(HTTPStatus.NOT_FOUND),
+    Errors.AlreadyExists: _CONFLICT,
+    Errors.PermissionDenied: ExtendedHTTPStatus.from_http_status(HTTPStatus.FORBIDDEN),
+    Errors.ResourceExhausted: ExtendedHTTPStatus.from_http_status(
+        HTTPStatus.TOO_MANY_REQUESTS
+    ),
+    Errors.FailedPrecondition: _BAD_REQUEST,
+    Errors.Aborted: _CONFLICT,
+    Errors.OutOfRange: _BAD_REQUEST,
+    Errors.Unimplemented: ExtendedHTTPStatus.from_http_status(
+        HTTPStatus.NOT_IMPLEMENTED
+    ),
+    Errors.Internal: _INTERNAL_SERVER_ERROR,
+    Errors.Unavailable: ExtendedHTTPStatus.from_http_status(
+        HTTPStatus.SERVICE_UNAVAILABLE
+    ),
+    Errors.DataLoss: _INTERNAL_SERVER_ERROR,
+    Errors.Unauthenticated: ExtendedHTTPStatus.from_http_status(
+        HTTPStatus.UNAUTHORIZED
+    ),
+    # Custom error codes not defined by Connec
+    Errors.NoError: ExtendedHTTPStatus.from_http_status(HTTPStatus.OK),
+    Errors.BadRoute: ExtendedHTTPStatus.from_http_status(HTTPStatus.NOT_FOUND),
+    Errors.Malformed: _BAD_REQUEST,
+}
+
+
+_http_status_code_to_error = {
+    400: Errors.Internal,
+    401: Errors.Unauthenticated,
+    403: Errors.PermissionDenied,
+    404: Errors.Unimplemented,
+    429: Errors.Unavailable,
+    502: Errors.Unavailable,
+    503: Errors.Unavailable,
+    504: Errors.Unavailable,
+}
+
+
+@dataclass(frozen=True, kw_only=True)
+class ConnectWireError:
```

Comment:
I have tried to follow the data structure from connect-go to make it easy to port from it